### PR TITLE
Fix stale cache protected length on waiting rematch

### PR DESCRIPTION
**Pull request by Kare0638**
## Motivation

While investigating https://github.com/sgl-project/sglang/issues/23392, a deterministic KV cache corruption reproducer, I found that waiting requests can have stale `cache_protected_len` after prefix rematching.

`SchedulePolicy._compute_prefix_matches()` refreshes `prefix_indices` and `last_node` for requests in the waiting queue. However, it did not refresh `cache_protected_len`. If another request inserts a longer shared prefix while a request is waiting, the waiting request can be rematched to the longer prefix while still keeping the old shorter protected length.

Later radix cache release paths use `cache_protected_len` as the boundary between shared radix KV and request-owned KV. A stale value can cause newly shared KV to be treated as request-owned overlap.

This PR fixes that metadata inconsistency. It is a partial fix for one concrete issue found during the investigation; a separate radix KV ownership issue may still remain.

### Concrete Example

Consider a waiting request `R`:

```text
R.input_ids = [1, 2, 3, 4]
R.prefix_indices = []
R.cache_protected_len = 0
R.req_pool_idx = 42
```

While `R` is still waiting, another request inserts the same prefix into the radix cache:

```text
radix tree: [1, 2, 3, 4] -> [kv0, kv1, kv2, kv3]
```

Later, `_compute_prefix_matches()` rematches `R` and finds this longer cached prefix:

```text
match_result.device_indices = [kv0, kv1, kv2, kv3]
match_result.last_device_node = node_1234
```

Before this PR, `_compute_prefix_matches()` refreshed only part of the request state:

```text
R.prefix_indices = [kv0, kv1, kv2, kv3]
R.last_node = node_1234
R.cache_protected_len = 0  # stale
```

This is inconsistent. `prefix_indices` and `last_node` say that `R` now uses a 4-token shared radix prefix, but `cache_protected_len` still says no KV is protected by the radix tree.

If `R` is released before it is scheduled again, cache release can compute the free range from the stale protected length:

```text
free(kv_indices[R.cache_protected_len : key_len])
= free(kv_indices[0 : 4])
```

That can free KV slots that are still referenced by the radix tree.

After this PR, the rematch also refreshes the protected boundary:

```text
R.prefix_indices = [kv0, kv1, kv2, kv3]
R.last_node = node_1234
R.cache_protected_len = 4
```

The same release path then computes:

```text
free(kv_indices[4 : 4])
```

so the shared radix prefix is not released as request-owned KV.

## Modifications

- Update `cache_protected_len` after waiting queue prefix rematch in `_compute_prefix_matches()`.
- Use `match_result.cache_protected_len` when available, with a fallback to `len(prefix_indices)`.
- Add a unit test covering the rematch path and verifying that `cache_protected_len` is refreshed together with `prefix_indices`.

## Accuracy Tests

This PR does not change model forward computation, kernels, sampling, or logits. I validated the metadata fix with a deterministic KV cache corruption reproducer from https://github.com/sgl-project/sglang/issues/23392:

- Repro script: https://gist.github.com/Yunzez/7dedd70360e9d2fdec76d72aa15b84fd#file-repro-py
- Archived finding JSON: https://gist.github.com/Yunzez/ba93887abcfcfae5f4e92b96330b6266
- GPU: A100 SXM
- Model: `Qwen/Qwen2.5-0.5B-Instruct`
- Temperature: `0.0`
- Speculative decoding: NGRAM
- `--speculative-num-draft-tokens 5`
- `--max-running-requests 8`
- Repeated trace replays: 15 to 60 runs

Before this fix, the reproducer could trigger all expected divergent requests:

```text
[DIVERGED] 4 request(s)

  6327450a (unexpected)
    variant 1: ' u1252_882 u'
    variant 2: ' u1252_888 u'

  77681d56 (expected)
    variant 1: ' u6212_144 u'
    variant 2: ' u6212_126 u'

  scn_0_retry (expected)
    variant 1: ' u7257_148 u72'
    variant 2: ' u7257_144 u72'

  scn_1_retry (expected)
    variant 1: ' u7257_144 u72'
    variant 2: ' u7257_148 u72'
    variant 3: ' u7257_1446 u7'

[STABLE] 2 request(s): 5c3e7afe, 622af575

RESULT: CONFIRMED (3/3 expected requests diverged)
```

After updating `cache_protected_len` on waiting queue rematch, the same reproducer often reduced to a single remaining expected divergence:

```text
[DIVERGED] 1 request(s)

  scn_0_retry (expected)
    variant 1: ' u7257_148 u72'
    variant 2: ' u7257_1446 u7'

[STABLE] 5 request(s): 5c3e7afe, 622af575, 6327450a, 77681d56, scn_1_retry

RESULT: PARTIAL (found=['scn_0_retry'], expected=['77681d56', 'scn_0_retry', 'scn_1_retry'])
```

Another run showed:

```text
[DIVERGED] 1 request(s)

  scn_1_retry (expected)
    variant 1: ' u7257_1446 u7'
    variant 2: ' u7257_148 u72'

[STABLE] 6 request(s): 5c3e7afe, 622af575, 6327450a, 77681d56, scn_0_retry, scn_2

RESULT: PARTIAL (found=['scn_1_retry'], expected=['77681d56', 'scn_0_retry', 'scn_1_retry'])
```

With radix cache disabled, the reproducer was stable:

```text
[CLEAN] no output divergence detected

[STABLE] 6 request(s): 5c3e7afe, 622af575, 6327450a, 77681d56, scn_0_retry, scn_1_retry

RESULT: NOT_REPRODUCED
```

This suggests the stale `cache_protected_len` fix removes part of the corruption surface, but the full stress reproducer can still expose a separate radix KV ownership issue. This PR is intentionally scoped as a partial correctness fix for the stale rematch metadata bug.

## Speed Tests and Profiling

Not applicable. This is a small scheduling metadata update and is not expected to affect inference speed.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). Not applicable; no user-facing behavior or API changed.
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). Not applicable; no model forward or performance path changed.
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Comment by Kare0638** [comment]
Additional context on why the archived repro can trigger this bug:

The repro script reconstructs prompts from the archived finding by combining a shared `pre{i}` prefix with request-specific suffix tokens. This creates multiple concurrent requests with long shared prefixes but different suffixes, for example the `u5008`, `u1252`, and `u4307` prompt families. In the investigation, those families had a real tokenizer LCP of `915`, so matching a 915-token radix prefix is valid.

The repro also replays the original fuzzer timing with async HTTP requests and client-side `cancel` / `disconnect` events. That creates the relevant window:

1. A request enters the waiting queue with an older/shorter matched prefix, e.g. `cache_protected_len = 402`.
2. Another request runs and inserts a longer shared prefix into the radix cache, e.g. real LCP `915`.
3. `_compute_prefix_matches()` rematches the waiting request and refreshes `prefix_indices` / `last_node` to the longer radix prefix.
4. Before this PR, `cache_protected_len` was not refreshed, so the request could be left in an inconsistent state:

```text
prefix_indices       = 915-token shared prefix
last_node            = radix node for that prefix
cache_protected_len  = 402  # stale
```
If the request is released before it is scheduled again, later cache release code can use the stale protected boundary and free KV in [402:915], even though that range is now part of the shared radix prefix. That makes subsequent deterministic temperature=0 requests able to diverge after legally hitting the radix prefix.

This PR fixes that stale metadata state by updating cache_protected_len together with prefix_indices and last_node during waiting queue prefix rematch.

**Comment by ErenAta16** [comment]
Traced this through `MatchResult` and the surrounding scheduler code before trusting the diff. `cache_protected_len` is a real field on `MatchResult` (`base_prefix_cache.py`, defaults to `None`, just missing from the class docstring), so this isn't referencing something that doesn't exist.

Checked the in-batch-prefix-caching block right below the insertion point too, since it reassigns a local `match_result` from a different tree (`self.waiting_queue_radix_tree`) for the small-prefix threshold check. That second `match_result` only feeds `in_batch_matching_prefixes`/`temporary_deprioritized` and an `insert()` call, it never touches `r.cache_protected_len` or `r.prefix_indices`, so there's no conflict between that block and this fix, they're using the match result from `match_prefix_for_req` for different purposes.

The fallback (`len(r.prefix_indices)` when `match_result.cache_protected_len` is `None`) only makes sense if `r.prefix_indices` has already been refreshed to the rematched value by that point, which lines up with the diff landing right after the `match_prefix_for_req` call that does that refresh.

Reasoning in the PR description matches the code: `_compute_prefix_matches` did update `prefix_indices`/`last_node` but not `cache_protected_len`, and downstream cache-release code treating a stale (too-short) protected length as the shared/owned boundary is exactly the kind of bug that'd silently free KV that's still shared. Test covers the direct case (insert into tree cache, then confirm rematch updates the protected length too). Looks correct.

**Comment by hnyls2002** [comment]
Closing as fixed on main. The cache_protected_len refresh on waiting rematch is now in schedule_batch.py:1388-1391 and schedule_policy.py:195-196 (landed in d6032c04b6). @Kare0638 please reopen if you still see the rematch-stale pattern on current main.

**Comment by ErenAta16** [comment]
Before this settles as closed, I went to look at the two locations, because both guards depend on a cache populating `MatchResult.cache_protected_len` and I could not find one that does.

The field defaults to `None` (`base_prefix_cache.py:208`). Searching every cache implementation for an assignment of it onto a `MatchResult`, the only two `cache_protected_len=` hits in the tree are f-string debug formatting:

```
mamba_radix_cache.py:784     f"{req.cache_protected_len=}, ..."
unified_radix_cache.py:900   f"{req.cache_protected_len=}, ..."
```

Every other reference is to `req.cache_protected_len`, either read in a release path or assigned in the cache-write path (`radix_cache.py:569`, `swa_radix_cache.py:565`, `mamba_radix_cache.py:802`, `unified_radix_cache.py:932`). `RadixCache.match_prefix` builds its result with four fields and no protected length (`radix_cache.py:430-435`, empty result at `365-374`), and `hiradix_cache.py` and `radix_cache_cpp.py` contain no reference to the field at all.

If that is right, `match_result.cache_protected_len` is `None` on every path today, and the two sites you cite then behave differently.

`schedule_batch.py:1388-1391` always takes the `else`, so it does refresh:

```python
if match_result.cache_protected_len is not None:
    self.cache_protected_len = match_result.cache_protected_len
else:
    self.cache_protected_len = len(self.prefix_indices)
```

`schedule_policy.py:195-196` has no `else`, so it never refreshes:

```python
if match_result.cache_protected_len is not None:
    req.cache_protected_len = match_result.cache_protected_len
```

That second one is the waiting-rematch site. `match_prefix_for_req` refreshes `req.prefix_indices` unconditionally about ten lines above it, and `_compute_prefix_matches` calls it for every request in the waiting queue, so a waiting request can still leave a rematch with `prefix_indices` advanced and `cache_protected_len` at its previous value, which is the state this PR was describing.

There are two honest readings. Either the field is forward-looking for a cache that will publish it and the working mechanism is the `schedule_batch` fallback, in which case the policy site likely wants the same `else`. Or something re-syncs the request before any release path reads the two together, in which case the `schedule_policy` half is not load-bearing and closing this is right.

I could not run it. There is no GPU on this machine, so the above comes from reading the files on current `main` rather than from a repro, and the sequencing question is exactly the part I cannot settle that way. Raising it only because the policy-side guard looks like it was meant to be doing work. @hnyls2002

**Comment by Kare0638** [comment]
Thanks @hnyls2002. Just one correction on the reference:
`d6032c04b6` is a docs-only commit (`[docs] Fix V4 Pro balanced recipe (#26451)`). The code at `schedule_policy.py:195-196` came from the refactor in #19746 (`5b7ce417d05`).

**Comment by ErenAta16** [comment]
Confirming @Kare0638's correction independently, and it turns out to matter more than a citation fix.

`d6032c04b6` is docs-only. Two files, both under `docs_new/`:

```
docs_new/cookbook/autoregressive/DeepSeek/DeepSeek-V4.mdx        +2/-2
docs_new/src/snippets/autoregressive/deepseek-v4-deployment.jsx  +4/-10
```

And `5b7ce417d05` (#19746, "[P/D disagg] - support decode side radix cache", 2026-05-01) is where `schedule_policy.py` got the guard:

```
python/sglang/srt/managers/schedule_policy.py  +37/-17
  + if match_result.cache_protected_len is not None:
  +     req.cache_protected_len = match_result.cache_protected_len

python/sglang/srt/managers/schedule_batch.py   +3/-1
  + self.cache_protected_len = 0
```

So those two lines are not a fix for this issue that landed later. They arrived in May as part of a P/D disaggregation feature, months before this PR, and `schedule_batch.py` got only the field initialiser in the same commit, not the guard with its `else`.

That changes what the close-out means. The waiting-rematch site was not fixed and then cited; it has been there since May, and since no cache implementation ever populates `MatchResult.cache_protected_len`, it has never executed. Whatever protection exists today comes entirely from the `else` fallback in `schedule_batch.py`, which is a different call path.

Still not something I can run, so I am not claiming a live repro. But the provenance is checkable and it removes the reading where the policy-side guard was a deliberate, working fix.
